File: frontend/src/api/catalog.ts
import { apiRequest } from "./client";

export type SeasonalityPublic = {
  id: number;
  dish_id: number;
  seasonality_mode: string;
  preferred_months: number[];
  allowed_months: number[];
  excluded_months: number[];
  seasonality_strength: string;
  created_at: string;
  updated_at: string;
};

export type Dish = {
  id: number;
  public_key: string;
  name: string;
  description: string | null;
  default_servings: number | null;
  default_prep_time_minutes: number | null;
  default_cook_time_minutes: number | null;
  default_difficulty: "easy" | "medium" | "hard" | null;
  course: "starter" | "main" | "dessert" | null;
  status: "draft" | "active" | "archived";
  image_url: string | null;
  suitable_for_lunch: boolean | null;
  suitable_for_dinner: boolean | null;
  weekday_friendly: boolean | null;
  leftovers_possible: boolean | null;
  freezer_friendly: boolean | null;
  kids_friendly: boolean | null;
  thermomix_possible: boolean | null;
  active: boolean;
  notes: string | null;
  created_at: string;
  updated_at: string;
  tag_ids: number[];
  computed_traits_json: Record<string, unknown> | null;
  seasonality: SeasonalityPublic | null;
};

export type SeasonalityInput = {
  seasonality_mode?: string;
  preferred_months?: number[];
  allowed_months?: number[];
  excluded_months?: number[];
  seasonality_strength?: string;
};

export type Tag = {
  id: number;
  name: string;
  family: string;
  description: string | null;
};

export type Recipe = {
  id: number;
  dish_id: number;
  public_key: string;
  sequence_number: number;
  variant_name: string;
  description: string | null;
  recipe_type: "standard" | "thermomix" | "other_appliance";
  is_main: boolean;
  is_thermomix: boolean;
  thermomix_model: string | null;
  source_url: string | null;
  servings: number | null;
  prep_time_minutes: number | null;
  cook_time_minutes: number | null;
  difficulty: "easy" | "medium" | "hard" | null;
  computed_traits_json: Record<string, unknown> | null;
  notes: string | null;
};

export type RecipeStep = {
  id: number;
  recipe_id: number;
  step_number: number;
  instruction: string;
  duration_seconds: number | null;
  temperature: string | null;
  timer_seconds: number | null;
  is_thermomix_step: boolean;
};

export type Ingredient = {
  id: number;
  canonical_name: string;
  display_name: string;
  category: string | null;
  food_group: string | null;
  family: string | null;
  default_unit_id: number | null;
  default_dimension: "mass" | "volume" | "count" | null;
  preferred_shopping_unit_id: number | null;
  aggregation_unit_id: number | null;
  aggregation_strategy:
    | "strict_same_dimension"
    | "prefer_mass"
    | "prefer_volume"
    | "prefer_count"
    | "allow_approximate_conversion"
    | "never_convert_count"
    | null;
  pantry_item: boolean;
  season_start_month: number | null;
  season_end_month: number | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
};

export type IngredientAlias = {
  id: number;
  ingredient_id: number;
  alias: string;
  language: string | null;
  created_at: string;
  updated_at: string;
};

export type IngredientConversion = {
  id: number;
  ingredient_id: number;
  from_unit_id: number;
  to_unit_id: number;
  from_unit_symbol: string;
  to_unit_symbol: string;
  factor: string;
  confidence: string;
  notes: string | null;
  approved: boolean;
  source: string | null;
  created_at: string;
  updated_at: string;
};

export type IngredientDetail = Ingredient & {
  aliases: IngredientAlias[];
  unit_conversions: IngredientConversion[];
};

export type IngredientInput = {
  canonical_name?: string;
  display_name: string;
  category?: string | null;
  food_group?: string | null;
  family?: string | null;
  default_unit_id?: number | null;
  default_dimension?: Ingredient["default_dimension"];
  preferred_shopping_unit_id?: number | null;
  aggregation_unit_id?: number | null;
  aggregation_strategy?: Ingredient["aggregation_strategy"];
  pantry_item?: boolean;
  season_start_month?: number | null;
  season_end_month?: number | null;
  notes?: string | null;
  aliases?: string[];
};

export type RecipeIngredient = {
  id: number;
  recipe_id: number;
  ingredient_id: number;
  quantity: string | null;
  unit_id: number | null;
  optional: boolean;
  notes: string | null;
};

export type Unit = {
  id: number;
  name: string;
  symbol: string;
  dimension: "mass" | "volume" | "count";
};

export type IngredientCategory = {
  id: string;
  label: string;
};

export type DishInput = {
  name: string;
  description?: string | null;
  image_url?: string | null;
  course?: Dish["course"];
  status?: Dish["status"];
  suitable_for_lunch?: boolean | null;
  suitable_for_dinner?: boolean | null;
  weekday_friendly?: boolean | null;
  leftovers_possible?: boolean | null;
  freezer_friendly?: boolean | null;
  kids_friendly?: boolean | null;
  notes?: string | null;
  tag_ids?: number[];
  seasonality?: SeasonalityInput | null;
};

export type IngredientResolveResponse = {
  status: "exact" | "suggestions" | "none";
  ingredient?: Ingredient;
  suggestions?: Ingredient[];
};

function withToken(token: string) {
  return { token };
}

export async function fetchDishes(token: string, activeOnly = false): Promise<Dish[]> {
  const query = activeOnly ? "?active_only=true" : "";
  return apiRequest<Dish[]>(`/api/dishes${query}`, withToken(token));
}

export async function fetchDish(token: string, dishId: number): Promise<Dish> {
  return apiRequest<Dish>(`/api/dishes/${dishId}`, withToken(token));
}

export async function createDish(token: string, payload: DishInput): Promise<Dish> {
  return apiRequest<Dish>("/api/dishes", {
    method: "POST",
    body: JSON.stringify(payload),
    ...withToken(token),
  });
}

export async function updateDish(token: string, dishId: number, payload: Partial<DishInput>): Promise<Dish> {
  return apiRequest<Dish>(`/api/dishes/${dishId}`, {
    method: "PUT",
    body: JSON.stringify(payload),
    ...withToken(token),
  });
}

export async function deleteDish(token: string, dishId: number): Promise<void> {
  return apiRequest<void>(`/api/dishes/${dishId}`, {
    method: "DELETE",
    ...withToken(token),
  });
}

export async function fetchTags(token: string): Promise<Tag[]> {
  return apiRequest<Tag[]>("/api/tags", withToken(token));
}

export async function fetchRecipes(token: string, dishId: number): Promise<Recipe[]> {
  return apiRequest<Recipe[]>(`/api/dishes/${dishId}/recipes`, withToken(token));
}

export async function createRecipe(
  token: string,
  dishId: number,
  payload: {
    variant_name: string;
    description?: string | null;
    recipe_type?: Recipe["recipe_type"];
    is_main?: boolean;
    servings?: number | null;
    prep_time_minutes?: number | null;
    cook_time_minutes?: number | null;
    difficulty?: Recipe["difficulty"];
    source_url?: string | null;
    notes?: string | null;
  },
): Promise<Recipe> {
  return apiRequest<Recipe>(`/api/dishes/${dishId}/recipes`, {
    method: "POST",
    body: JSON.stringify(payload),
    ...withToken(token),
  });
}

export async function updateRecipe(
  token: string,
  recipeId: number,
  payload: {
    variant_name?: string;
    description?: string | null;
    recipe_type?: Recipe["recipe_type"];
    is_main?: boolean;
    servings?: number | null;
    prep_time_minutes?: number | null;
    cook_time_minutes?: number | null;
    difficulty?: Recipe["difficulty"];
    source_url?: string | null;
    notes?: string | null;
  },
): Promise<Recipe> {
  return apiRequest<Recipe>(`/api/recipes/${recipeId}`, {
    method: "PUT",
    body: JSON.stringify(payload),
    ...withToken(token),
  });
}

export async function fetchRecipe(token: string, recipeId: number): Promise<Recipe> {
  return apiRequest<Recipe>(`/api/recipes/${recipeId}`, withToken(token));
}

export async function deleteRecipe(token: string, recipeId: number): Promise<void> {
  return apiRequest<void>(`/api/recipes/${recipeId}`, {
    method: "DELETE",
    ...withToken(token),
  });
}

export async function fetchRecipeSteps(token: string, recipeId: number): Promise<RecipeStep[]> {
  return apiRequest<RecipeStep[]>(`/api/recipes/${recipeId}/steps`, withToken(token));
}

export async function createRecipeStep(
  token: string,
  recipeId: number,
  payload: { step_number: number; instruction: string },
): Promise<RecipeStep> {
  return apiRequest<RecipeStep>(`/api/recipes/${recipeId}/steps`, {
    method: "POST",
    body: JSON.stringify(payload),
    ...withToken(token),
  });
}

export async function updateRecipeStep(
  token: string,
  stepId: number,
  payload: { step_number?: number; instruction?: string },
): Promise<RecipeStep> {
  return apiRequest<RecipeStep>(`/api/recipe-steps/${stepId}`, {
    method: "PUT",
    body: JSON.stringify(payload),
    ...withToken(token),
  });
}

export async function deleteRecipeStep(token: string, stepId: number): Promise<void> {
  return apiRequest<void>(`/api/recipe-steps/${stepId}`, {
    method: "DELETE",
    ...withToken(token),
  });
}

export async function fetchRecipeIngredients(token: string, recipeId: number): Promise<RecipeIngredient[]> {
  return apiRequest<RecipeIngredient[]>(`/api/recipes/${recipeId}/ingredients`, withToken(token));
}

export async function addRecipeIngredient(
  token: string,
  recipeId: number,
  payload: {
    ingredient_id?: number;
    proposed_name?: string;
    quantity?: string | null;
    unit_id?: number | null;
    optional?: boolean;
  },
): Promise<RecipeIngredient> {
  return apiRequest<RecipeIngredient>(`/api/recipes/${recipeId}/ingredients`, {
    method: "POST",
    body: JSON.stringify(payload),
    ...withToken(token),
  });
}

export async function updateRecipeIngredient(
  token: string,
  itemId: number,
  payload: {
    quantity?: string | null;
    unit_id?: number | null;
    optional?: boolean;
    notes?: string | null;
  },
): Promise<RecipeIngredient> {
  return apiRequest<RecipeIngredient>(`/api/recipe-ingredients/${itemId}`, {
    method: "PUT",
    body: JSON.stringify(payload),
    ...withToken(token),
  });
}

export async function deleteRecipeIngredient(token: string, itemId: number): Promise<void> {
  return apiRequest<void>(`/api/recipe-ingredients/${itemId}`, {
    method: "DELETE",
    ...withToken(token),
  });
}

export async function fetchIngredients(token: string, search?: string): Promise<Ingredient[]> {
  const query = search?.trim() ? `?search=${encodeURIComponent(search.trim())}` : "";
  return apiRequest<Ingredient[]>(`/api/ingredients${query}`, withToken(token));
}

export async function fetchIngredient(token: string, ingredientId: number): Promise<IngredientDetail> {
  return apiRequest<IngredientDetail>(`/api/ingredients/${ingredientId}`, withToken(token));
}

export async function createIngredient(token: string, payload: IngredientInput): Promise<Ingredient> {
  return apiRequest<Ingredient>("/api/ingredients", {
    method: "POST",
    body: JSON.stringify(payload),
    ...withToken(token),
  });
}

export async function updateIngredient(
  token: string,
  ingredientId: number,
  payload: Partial<IngredientInput>,
): Promise<Ingredient> {
  return apiRequest<Ingredient>(`/api/ingredients/${ingredientId}`, {
    method: "PUT",
    body: JSON.stringify(payload),
    ...withToken(token),
  });
}

export async function deleteIngredient(token: string, ingredientId: number): Promise<void> {
  return apiRequest<void>(`/api/ingredients/${ingredientId}`, {
    method: "DELETE",
    ...withToken(token),
  });
}

export async function createIngredientAlias(
  token: string,
  ingredientId: number,
  payload: { alias: string; language?: string | null },
): Promise<IngredientAlias> {
  return apiRequest<IngredientAlias>(`/api/ingredients/${ingredientId}/aliases`, {
    method: "POST",
    body: JSON.stringify(payload),
    ...withToken(token),
  });
}

export async function deleteIngredientAlias(token: string, aliasId: number): Promise<void> {
  return apiRequest<void>(`/api/ingredient-aliases/${aliasId}`, {
    method: "DELETE",
    ...withToken(token),
  });
}

export async function createIngredientConversion(
  token: string,
  ingredientId: number,
  payload: {
    from_unit_id: number;
    to_unit_id: number;
    factor: string;
    confidence?: string;
    notes?: string | null;
    approved?: boolean;
    source?: string;
  },
): Promise<IngredientConversion> {
  return apiRequest<IngredientConversion>(`/api/ingredients/${ingredientId}/conversions`, {
    method: "POST",
    body: JSON.stringify(payload),
    ...withToken(token),
  });
}

export async function updateIngredientConversion(
  token: string,
  conversionId: number,
  payload: {
    factor?: string;
    confidence?: string;
    notes?: string | null;
    approved?: boolean;
    source?: string;
  },
): Promise<IngredientConversion> {
  return apiRequest<IngredientConversion>(`/api/ingredient-conversions/${conversionId}`, {
    method: "PUT",
    body: JSON.stringify(payload),
    ...withToken(token),
  });
}

export async function deleteIngredientConversion(token: string, conversionId: number): Promise<void> {
  return apiRequest<void>(`/api/ingredient-conversions/${conversionId}`, {
    method: "DELETE",
    ...withToken(token),
  });
}

export async function resolveIngredient(token: string, proposedName: string): Promise<IngredientResolveResponse> {
  return apiRequest<IngredientResolveResponse>("/api/ingredients/resolve", {
    method: "POST",
    body: JSON.stringify({ proposed_name: proposedName }),
    ...withToken(token),
  });
}

export async function confirmIngredient(
  token: string,
  payload: {
    action: "create" | "map" | "alias";
    proposed_name: string;
    display_name?: string;
    ingredient_id?: number;
  },
): Promise<Ingredient> {
  return apiRequest<Ingredient>("/api/ingredients/confirm", {
    method: "POST",
    body: JSON.stringify(payload),
    ...withToken(token),
  });
}

export async function fetchUnits(token: string): Promise<Unit[]> {
  return apiRequest<Unit[]>("/api/units", withToken(token));
}

export async function fetchIngredientCategories(token: string): Promise<IngredientCategory[]> {
  return apiRequest<IngredientCategory[]>("/api/ingredient-categories", withToken(token));
}

File: backend/mealroulette/schemas/catalog.py
from datetime import datetime
from decimal import Decimal
from enum import Enum

from pydantic import BaseModel, ConfigDict, Field, model_validator

from mealroulette.models.enums import (
    AggregationStrategy,
    ConversionConfidence,
    ConversionSource,
    DifficultyLevel,
    DishCourse,
    DishStatus,
    RecipeType,
    SeasonalityMode,
    SeasonalityStrength,
    ServingTemperature,
    UnitDimension,
    VegetableLevel,
)


class UnitPublic(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    symbol: str
    dimension: UnitDimension
    conversion_to_base: Decimal
    created_at: datetime
    updated_at: datetime


class TagPublic(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    family: str
    description: str | None
    created_at: datetime
    updated_at: datetime


class TagCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=64)
    family: str = Field(min_length=1, max_length=64)
    description: str | None = None


class TagUpdateRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=64)
    family: str | None = Field(default=None, min_length=1, max_length=64)
    description: str | None = None


class IngredientCategoryPublic(BaseModel):
    id: str
    label: str


class IngredientPublic(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    canonical_name: str
    display_name: str
    category: str | None
    food_group: str | None
    storage_class: str | None = None
    storage_after_opening: str | None = None
    culinary_category: str | None = None
    product_form: str | None = None
    preservation: str | None = None
    traits_json: dict | None = None
    family: str | None
    default_unit_id: int | None
    default_dimension: UnitDimension | None
    preferred_shopping_unit_id: int | None
    aggregation_unit_id: int | None
    aggregation_strategy: AggregationStrategy | None
    pantry_item: bool
    season_start_month: int | None
    season_end_month: int | None
    notes: str | None
    created_at: datetime
    updated_at: datetime


class IngredientUnitConversionPublic(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    ingredient_id: int
    from_unit_id: int
    to_unit_id: int
    from_unit_symbol: str
    to_unit_symbol: str
    factor: Decimal
    confidence: ConversionConfidence
    notes: str | None
    approved: bool
    source: ConversionSource | None
    created_at: datetime
    updated_at: datetime


class IngredientDetailPublic(IngredientPublic):
    aliases: list["IngredientAliasPublic"] = Field(default_factory=list)
    unit_conversions: list[IngredientUnitConversionPublic] = Field(default_factory=list)


class IngredientCreateRequest(BaseModel):
    canonical_name: str = Field(min_length=1, max_length=128)
    display_name: str = Field(min_length=1, max_length=128)
    category: str | None = Field(default=None, max_length=64)
    food_group: str | None = Field(default=None, max_length=64)
    storage_class: str | None = Field(default=None, max_length=32)
    culinary_category: str | None = Field(default=None, max_length=64)
    product_form: str | None = Field(default=None, max_length=32)
    preservation: str | None = Field(default=None, max_length=32)
    family: str | None = Field(default=None, max_length=64)
    default_unit_id: int | None = None
    default_dimension: UnitDimension | None = None
    preferred_shopping_unit_id: int | None = None
    aggregation_unit_id: int | None = None
    aggregation_strategy: AggregationStrategy | None = None
    pantry_item: bool = False
    season_start_month: int | None = Field(default=None, ge=1, le=12)
    season_end_month: int | None = Field(default=None, ge=1, le=12)
    notes: str | None = None
    aliases: list[str] = Field(default_factory=list)


class IngredientUpdateRequest(BaseModel):
    display_name: str | None = Field(default=None, min_length=1, max_length=128)
    category: str | None = Field(default=None, max_length=64)
    food_group: str | None = Field(default=None, max_length=64)
    storage_class: str | None = Field(default=None, max_length=32)
    culinary_category: str | None = Field(default=None, max_length=64)
    product_form: str | None = Field(default=None, max_length=32)
    preservation: str | None = Field(default=None, max_length=32)
    family: str | None = Field(default=None, max_length=64)
    default_unit_id: int | None = None
    default_dimension: UnitDimension | None = None
    preferred_shopping_unit_id: int | None = None
    aggregation_unit_id: int | None = None
    aggregation_strategy: AggregationStrategy | None = None
    pantry_item: bool | None = None
    season_start_month: int | None = Field(default=None, ge=1, le=12)
    season_end_month: int | None = Field(default=None, ge=1, le=12)
    notes: str | None = None


class IngredientUnitConversionCreateRequest(BaseModel):
    from_unit_id: int
    to_unit_id: int
    factor: Decimal = Field(gt=0)
    confidence: ConversionConfidence = ConversionConfidence.medium
    notes: str | None = None
    approved: bool = False
    source: ConversionSource = ConversionSource.manual


class IngredientUnitConversionUpdateRequest(BaseModel):
    factor: Decimal | None = Field(default=None, gt=0)
    confidence: ConversionConfidence | None = None
    notes: str | None = None
    approved: bool | None = None
    source: ConversionSource | None = None

    @model_validator(mode="before")
    @classmethod
    def reject_null_non_nullable_fields(cls, data: object) -> object:
        if not isinstance(data, dict):
            return data
        for field in ("factor", "confidence", "approved"):
            if field in data and data[field] is None:
                raise ValueError(f"{field} cannot be null")
        return data


class IngredientAliasPublic(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    ingredient_id: int
    alias: str
    language: str | None
    created_at: datetime
    updated_at: datetime


class IngredientAliasCreateRequest(BaseModel):
    alias: str = Field(min_length=1, max_length=128)
    language: str | None = Field(default=None, max_length=16)


class IngredientResolveRequest(BaseModel):
    proposed_name: str = Field(min_length=1, max_length=128)


class IngredientResolveResponse(BaseModel):
    status: str
    query: str | None = None
    matched_on: str | None = None
    matched_value: str | None = None
    ingredient: IngredientPublic | None = None
    suggestions: list[IngredientPublic] = Field(default_factory=list)


class IngredientConfirmAction(str, Enum):
    create = "create"
    map = "map"
    alias = "alias"


class IngredientConfirmRequest(BaseModel):
    action: IngredientConfirmAction
    proposed_name: str = Field(min_length=1, max_length=128)
    ingredient_id: int | None = None
    display_name: str | None = Field(default=None, min_length=1, max_length=128)
    category: str | None = None
    default_unit_id: int | None = None
    language: str | None = None


class SeasonalityPublic(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    dish_id: int
    seasonality_mode: SeasonalityMode
    preferred_months: list[int]
    allowed_months: list[int]
    excluded_months: list[int]
    seasonality_strength: SeasonalityStrength
    created_at: datetime
    updated_at: datetime


class SeasonalityUpsertRequest(BaseModel):
    seasonality_mode: SeasonalityMode = SeasonalityMode.all_year
    preferred_months: list[int] = Field(default_factory=list)
    allowed_months: list[int] = Field(default_factory=list)
    excluded_months: list[int] = Field(default_factory=list)
    seasonality_strength: SeasonalityStrength = SeasonalityStrength.neutral


class RecipeStepPublic(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    recipe_id: int
    step_number: int
    instruction: str
    duration_seconds: int | None
    temperature: str | None
    timer_seconds: int | None
    is_thermomix_step: bool
    metadata_json: dict | None
    created_at: datetime
    updated_at: datetime


class RecipeStepCreateRequest(BaseModel):
    step_number: int = Field(ge=1)
    instruction: str = Field(min_length=1)
    duration_seconds: int | None = Field(default=None, ge=0)
    temperature: str | None = Field(default=None, max_length=64)
    timer_seconds: int | None = Field(default=None, ge=0)
    is_thermomix_step: bool = False
    metadata_json: dict | None = None


class RecipeStepUpdateRequest(BaseModel):
    step_number: int | None = Field(default=None, ge=1)
    instruction: str | None = Field(default=None, min_length=1)
    duration_seconds: int | None = Field(default=None, ge=0)
    temperature: str | None = Field(default=None, max_length=64)
    timer_seconds: int | None = Field(default=None, ge=0)
    is_thermomix_step: bool | None = None
    metadata_json: dict | None = None


class RecipeIngredientPublic(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    recipe_id: int
    ingredient_id: int
    quantity: Decimal | None
    unit_id: int | None
    optional: bool
    notes: str | None
    created_at: datetime
    updated_at: datetime


class RecipeIngredientCreateRequest(BaseModel):
    ingredient_id: int | None = None
    proposed_name: str | None = Field(default=None, min_length=1, max_length=128)
    quantity: Decimal | None = Field(default=None, ge=0)
    unit_id: int | None = None
    optional: bool = False
    notes: str | None = None


class RecipeIngredientUpdateRequest(BaseModel):
    quantity: Decimal | None = Field(default=None, ge=0)
    unit_id: int | None = None
    optional: bool | None = None
    notes: str | None = None


class RecipePublic(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    dish_id: int
    public_key: str
    sequence_number: int
    variant_name: str
    description: str | None
    recipe_type: RecipeType
    is_main: bool
    is_thermomix: bool
    thermomix_model: str | None
    source_url: str | None
    servings: int | None
    prep_time_minutes: int | None
    cook_time_minutes: int | None
    difficulty: DifficultyLevel | None
    computed_traits_json: dict | None
    notes: str | None
    created_at: datetime
    updated_at: datetime


class RecipeCreateRequest(BaseModel):
    variant_name: str = Field(min_length=1, max_length=128)
    description: str | None = None
    recipe_type: RecipeType | None = None
    is_main: bool | None = None
    is_thermomix: bool | None = None
    thermomix_model: str | None = Field(default=None, max_length=32)
    source_url: str | None = Field(default=None, max_length=512)
    servings: int | None = Field(default=None, ge=1)
    prep_time_minutes: int | None = Field(default=None, ge=0)
    cook_time_minutes: int | None = Field(default=None, ge=0)
    difficulty: DifficultyLevel | None = None
    notes: str | None = None


class RecipeUpdateRequest(BaseModel):
    variant_name: str | None = Field(default=None, min_length=1, max_length=128)
    description: str | None = None
    recipe_type: RecipeType | None = None
    is_main: bool | None = None
    is_thermomix: bool | None = None
    thermomix_model: str | None = Field(default=None, max_length=32)
    source_url: str | None = Field(default=None, max_length=512)
    servings: int | None = Field(default=None, ge=1)
    prep_time_minutes: int | None = Field(default=None, ge=0)
    cook_time_minutes: int | None = Field(default=None, ge=0)
    difficulty: DifficultyLevel | None = None
    notes: str | None = None


class DishPublic(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    public_key: str
    name: str
    description: str | None
    default_servings: int | None
    default_prep_time_minutes: int | None
    default_cook_time_minutes: int | None
    default_difficulty: DifficultyLevel | None
    course: DishCourse | None
    status: DishStatus
    image_url: str | None
    suitable_for_lunch: bool | None
    suitable_for_dinner: bool | None
    weekday_friendly: bool | None
    leftovers_possible: bool | None
    freezer_friendly: bool | None
    kids_friendly: bool | None
    thermomix_possible: bool | None
    active: bool
    notes: str | None
    created_at: datetime
    updated_at: datetime
    tag_ids: list[int] = Field(default_factory=list)
    computed_traits_json: dict | None = None
    seasonality: SeasonalityPublic | None = None


class DishDetailPublic(DishPublic):
    recipes: list[RecipePublic] = Field(default_factory=list)


class DishCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    description: str | None = None
    default_servings: int | None = Field(default=None, ge=1)
    default_prep_time_minutes: int | None = Field(default=None, ge=0)
    default_cook_time_minutes: int | None = Field(default=None, ge=0)
    default_difficulty: DifficultyLevel | None = None
    course: DishCourse | None = None
    status: DishStatus = DishStatus.active
    image_url: str | None = Field(default=None, max_length=512)
    suitable_for_lunch: bool | None = None
    suitable_for_dinner: bool | None = None
    weekday_friendly: bool | None = None
    leftovers_possible: bool | None = None
    freezer_friendly: bool | None = None
    kids_friendly: bool | None = None
    active: bool = True
    notes: str | None = None
    tag_ids: list[int] = Field(default_factory=list)
    seasonality: SeasonalityUpsertRequest | None = None


class DishUpdateRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None
    default_servings: int | None = Field(default=None, ge=1)
    default_prep_time_minutes: int | None = Field(default=None, ge=0)
    default_cook_time_minutes: int | None = Field(default=None, ge=0)
    default_difficulty: DifficultyLevel | None = None
    course: DishCourse | None = None
    status: DishStatus | None = None
    image_url: str | None = Field(default=None, max_length=512)
    suitable_for_lunch: bool | None = None
    suitable_for_dinner: bool | None = None
    weekday_friendly: bool | None = None
    leftovers_possible: bool | None = None
    freezer_friendly: bool | None = None
    kids_friendly: bool | None = None
    active: bool | None = None
    notes: str | None = None
    tag_ids: list[int] | None = None
    seasonality: SeasonalityUpsertRequest | None = None
